### PR TITLE
Semantic checks for constraints on types

### PR DESCRIPTION
**Pull request by psteinfeld**
I implemented and added tests for constraints C703, C704, C705, C706,
and C796.  In some cases, the code and/or test already existed, and all
I did was add a notation indicating the associated constraint.

### CODE REVIEW

File: test/semantics/resolve70.f90 [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  type :: extensible
+  end type
+  type, bind(c) :: inextensible
+  end type
+
+  ! This one's OK
+  class(extensible) :: y
+
+  !ERROR: A declaration that uses the CLASS keyword must specify an extensible derived type.  'inextensible' is not extensible
+  class(inextensible) :: x
```
Comment (tskeith) [inline test/semantics/resolve70.f90]: How about:
`Non-extensible derived type 'inextensible' may not be used with CLASS keyword

File: lib/semantics/check-declarations.cpp [hunk 2]
```diff
@@ -495,7 +495,7 @@ void CheckHelper::CheckDerivedType(
   }
   if (const DeclTypeSpec * parent{FindParentTypeSpec(symbol)}) {
     const DerivedTypeSpec *parentDerived{parent->AsDerived()};
-    if (!IsExtensibleType(parentDerived)) {
+    if (!IsExtensibleType(parentDerived)) {  // c705
```
Comment (tskeith) [inline lib/semantics/check-declarations.cpp]: c705 -> C705 for consistency.

File: lib/semantics/resolve-names.cpp [hunk 3]
```diff
@@ -3287,11 +3294,29 @@ bool DeclarationVisitor::Pre(const parser::DeclarationTypeSpec::Type &) {
...
+
 bool DeclarationVisitor::Pre(const parser::DeclarationTypeSpec::Class &) {
   SetDeclTypeSpecCategory(DeclTypeSpec::Category::ClassDerived);
   return true;
 }
 
+void DeclarationVisitor::Post(
+    const parser::DeclarationTypeSpec::Class &parsedClass) {
+  const auto &typeName{std::get<parser::Name>(parsedClass.derived.t)};
+  const std::optional<DerivedTypeSpec> spec{ResolveDerivedType(typeName)};
```
Comment (klausler) [inline lib/semantics/resolve-names.cpp]: Consider structuring this code with the declaration of `spec` (which need not be `const` and could be `auto') as the predicate of an `if` statement.

File: lib/semantics/resolve-names.cpp [hunk 4]
```diff
@@ -1612,6 +1613,12 @@ void DeclTypeSpecVisitor::Post(
 void DeclTypeSpecVisitor::MakeNumericType(TypeCategory category, int kind) {
   SetDeclTypeSpec(context().MakeNumericType(category, kind));
 }
+ 
+void DeclTypeSpecVisitor::CheckForAbstractType(const Symbol &typeSymbol) {
```
Comment (klausler) [inline lib/semantics/resolve-names.cpp]: If this function is only called once, I don't think it provides much value over its inline textual expansion.

File: test/semantics/resolve70.f90 (hunk 1 above)
Comment (psteinfeld) [reply test/semantics/resolve70.f90]: Good suggestion.  I'll change it.

File: lib/semantics/check-declarations.cpp (hunk 2 above)
Comment (psteinfeld) [reply lib/semantics/check-declarations.cpp]: My bad.  I'll change it.

File: lib/semantics/resolve-names.cpp (hunk 3 above)
Comment (psteinfeld) [reply lib/semantics/resolve-names.cpp]: Thanks, Peter.  Will do.

File: lib/semantics/resolve-names.cpp (hunk 4 above)
Comment (psteinfeld) [reply lib/semantics/resolve-names.cpp]: Agreed.  It's called twice, once in the `Post` function for `parser::TypeSpec` and again in the `Post` function for `parser::DeclarationTypeSpec`.  It just felt wrong to copy the code, so I created a separate function.

**Pull request merged**